Introducing binary search tree implementation

spirosmaggioros opened the pull request:
Implemented a binary serach tree in zig:
### Methods:
* Implemented methods for inserting (`insert`), removing (`remove`), and searching (`search`) elements in the tree.
* Added traversal methods for inorder (`inorder`), preorder (`preorder`), and postorder (`postorder`) traversals.
* Included a `destroy` method to deallocate memory used by the tree.

Also, the PR includes the unit tests for the container.

spirosmaggioros committed:
Introducing binary search tree implementation

spirosmaggioros committed:
Changed container name to match previous work

kassane commented:
Hi @spirosmaggioros,

Thanks for your contribution and happy new year!!

What is the difference compared to the previous implementation?

https://github.com/TheAlgorithms/Zig/blob/main/search%2FbinarySearchTree.zig

spirosmaggioros commented:
Happy new year to you too! 

Actually i want to contribute a lot and i started by trying fixing the already implemented data structures. 
My implementation has more methods(the ones that we both have are the same) and also they are more user friendly. Feels more like a class(accepts T key instead of nodes), thus, having an allocator as well. 

I can try and merge my additions into the already implemented file if you wish.

Thanks.
Spiros.

kassane commented:
> I can try and merge my additions into the already implemented file if you wish.

Please, it would help a lot, avoiding redundancies.

spirosmaggioros committed:
Updated Binary Search Tree data structure

Now the new BST function has more methods and in general feels more like my previous DoublyLinkedList implementation.

spirosmaggioros commented:
Just commited this. I updated the BinarySearchTree.zig with the new implementation instead of creating one at dataStructures. 

@kassane Do i have to add this to build.zig and runall.sh?

kassane commented:
> Do i have to add this to build.zig and runall.sh?

Again?

https://github.com/TheAlgorithms/Zig/blob/e68f833e53f176aaea76cffe97da3bd7b10add06/build.zig#L47-L53
https://github.com/TheAlgorithms/Zig/blob/e68f833e53f176aaea76cffe97da3bd7b10add06/runall.cmd#L34
https://github.com/TheAlgorithms/Zig/blob/e68f833e53f176aaea76cffe97da3bd7b10add06/runall.sh#L34

rebase and add [doublyLinkedList](https://github.com/TheAlgorithms/Zig/blob/main/dataStructures/doublyLinkedList.zig)

spirosmaggioros commented:
> > Do i have to add this to build.zig and runall.sh?
> 
> Again?
> 
> https://github.com/TheAlgorithms/Zig/blob/e68f833e53f176aaea76cffe97da3bd7b10add06/build.zig#L47-L53
> 
> 
> https://github.com/TheAlgorithms/Zig/blob/e68f833e53f176aaea76cffe97da3bd7b10add06/runall.cmd#L34
> 
> 
> https://github.com/TheAlgorithms/Zig/blob/e68f833e53f176aaea76cffe97da3bd7b10add06/runall.sh#L34
> 
> rebase and add [doublyLinkedList](https://github.com/TheAlgorithms/Zig/blob/main/dataStructures/doublyLinkedList.zig)

Sorry, i forgot it exists.

spirosmaggioros committed:
Merge branch 'TheAlgorithms:main' into main

kassane commented on search/binarySearchTree.zig:
```diff
@@ -1,82 +1,241 @@
-    const result = Tree(i32).search(tree.root, 3);
-    try expect(result == null);
-}
+test "Testing Binary Search Tree" {
```
Please! Break down or add more tests to evaluate the run-time of each operation. Like the previous example, following the zig-std tests format.

spirosmaggioros committed:
Added some more unit tests

spirosmaggioros committed:
Reformatting code

spirosmaggioros committed:
Merge branch 'main' into main

kassane commented:
Awesome contributions, @spirosmaggioros
